src/main/java/com/google/step/servlets/PromoCodeServlet.java

```diff
@@ -1,19 +1,56 @@
 package com.google.step.servlets;
 
+import com.google.step.youtube.YoutubeInfoScraper;
+import com.google.step.youtube.PromoCode;
+import com.google.gson.Gson;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import java.io.IOException;
+import java.util.List;
+import java.util.Optional;
 
 @WebServlet("/promo-code")
 public class PromoCodeServlet extends HttpServlet {
-    
+
+    private YoutubeInfoScraper infoScraper;
+
+    public PromoCodeServlet(YoutubeInfoScraper infoScraper) {
+        this.infoScraper = infoScraper;
+    }
+    public PromoCodeServlet() {
+        init();
+    }
+
     @Override
     public void init() {
+        infoScraper = new YoutubeInfoScraper();
     }
-
+    /**
+     * Takes in formInput as a parameter and then checks for a promoCodes based on the channelId.
+     * If the channelId is not proper then the json will return false.
+     * In a case where an exception is thrown the json will return false.
+     */
     @Override
     public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String channelId = request.getParameter("formInput");
+        response.setContentType("application/json");
+        try {
+            Optional<String> playlistId = infoScraper.scrapeChannelUploadPlaylist(channelId);
+            if (playlistId.isPresent()) {
+                Optional<List<PromoCode>> promoCodeList = infoScraper.scrapePromoCodesFromPlaylist(playlistId.get());
+                if (promoCodeList.isPresent()) {
+                    response.getWriter().println(new Gson().toJson(promoCodeList.get()));
+                } else {
+                    response.getWriter().println(new Gson().toJson(false));
```

Comment:
so you have a lot of if/else statements here.  There are several techniques to help with that.  Optiona.orElse will help, also consider maybe pulling most of doGet into a helper method:

`List<PromoCode> getPromoCodeList(String channelId)`

`doGet` will only own the parameter parsing and the Gson construction.

Then use guard clauses (see go/tott-455 or https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html) to immediately return if you encounter things that result in an empty result.  I.e.:
```
if (playlistId.isPresent()) {
return ImmutableList.of();
}

```

this might help get rid of some of the stuff here: 
if (xyz) {
// stuff
if (abc) { stuff } else {
// other stuff
}
//other other stuff
}

(Hard for me to see the end result unless i actually applied it.  kind requires some finessing and trying what looks most clean.)

If you need clarification over GVC, feel free to ping me and we can chat about it.